### ISSUE TITLE
Checking sentence attributes of Tokens of unit test Doc causes a segfault sometimes

### ISSUE THREAD

**Issue by swartchris8** [explosion/spaCy#5075]
I came across this issue while trying to write a complex spacy Doc based unit test. It seems checking the sent attribute of tokens of the doc created below causes a segfault sometimes. In my legacy spacy==2.0.12 dev env it always causes a segfault in the newer 2.2.3 only sometimes.

## How to reproduce the behaviour

Run below script 4-10 times and you should get a segfault:

```
from spacy.tests.util import get_doc
from spacy.vocab import Vocab

# "A simple method is used to compare the efficacy of ropinirole and rasagiline for Parkinson's disease."
words = ["A", "simple method"] + "is used to compare the efficacy of ropinirole and rasagiline for".split() + ["Parkinson's disease"]
vocab_words = """"A simple method is used to compare the efficacy of ropinirole and rasagiline for Parkinson's disease""".split()
vocab = Vocab(strings=vocab_words)
dependencies = [
    "det",  # A
    "nsubjpass",  # simple method
    "auxpass",  # is
    "ROOT",  # used
    "aux",  # to
    "xcomp",  # compare
    "det",  # the
    "dobj",  # efficacy
    "prep",  # of
    "pobj",  # ropinirole
    "cc",  # and
    "conj",  # rasagiline
    "prep",  # for
    "pobj",  # Parkinson's disease
]
tags = ["DT", "JJ", "VBZ", "VBN", "TO", "VB", "DT", "NN", "IN", "NN", "CC", "NN", "IN", "NNP"]
pos_tags = [
    "DET", "ADJ", "VERB", "VERB", "PART", "VERB", "DET", "NOUN", "ADP", "NOUN", "CCONJ", "NOUN", "ADP",
    "PROPN"]
heads = [1, 3, 3, 3, 5, 3, 7, 5, 7, 8, 9, 9, 9, 12]

en_doc = get_doc(vocab, words, pos_tags, heads, dependencies, tags)

for t in en_doc:
    print(t, t.sent)
```

## Your Environment

* **spaCy version:** 2.2.3
* **Platform:** Darwin-19.3.0-x86_64-i386-64bit
* **Python version:** 3.6.8
* Environment Information: Clean virtualenv with only spacy in it installed with `pip install spacy`

**Comment by adrianeboyd** [comment]
Thanks for the report! I strongly suspect that `get_doc()` is the source of the problem rather than `Token.sent`. You could try constructing your doc with `Doc` rather than `get_doc()` and see if you run into the same problems. A sketch:

```
from spacy.tokens import Doc

doc = Doc(vocab, words=words) # add `spaces` if you need it

for t, pos, tag, head, dep in zip(doc, pos_tags, tags, heads, dependencies):
    t.tag_ = tag
    t.pos_ = pos
    t.head = doc[head]
    t.dep_ = dep

doc.is_tagged = True
doc.is_parsed = True

for t in doc:
    print(t, t.sent)
```

`test.utils.get_doc()` is a bit buggy in various ways and @svlandeg has just been working on improving it in #5049, not due to segfaults directly (although I've seen the same kind of segfaults), and I think that her modifications should address these problems, too.

**Comment by svlandeg** [comment]
@adrianeboyd is probably right that `get_doc` is the culprit here. Unfortunately, it doesn't look like the changes from PR #5049 fix this particular issue :(

When I try to run the original code, I actually get into an infinite loop which is fixed with [this edit](https://github.com/svlandeg/spaCy/commit/42c6bacdf74fefadfd50d0a6953db0067b9a8e57) (which I think should have been there all along?). But after that small fix, and with the changes of PR #5049 merged, I do still run into a segfault :( So we'll have to revise `get_doc` further, I assume?

**Comment by adrianeboyd** [comment]
An infinite loop was not the intention there, no. I had made some local changes to `get_doc()` a while back to address segfaults, but I thought it mainly had to do with attribute arrays with differing lengths, and @svlandeg's code shouldn't allow that, so maybe I had just fixed it enough to get the tests to pass where tokenizer modifications cause problems and I hadn't identified the real issue. I will look into it again...

**Comment by adrianeboyd** [comment]
I suspect the problem is with the heads. `Doc.from_array()` expects relative heads rather than absolute heads. This should be detected and caught better.

**Comment by svlandeg** [comment]
Ah good point! Are you looking into this further or want me to?

**Comment by adrianeboyd** [comment]
Adjust the heads:

```
heads = [1, 3, 3, 3, 5, 3, 7, 5, 7, 8, 9, 9, 9, 12]
for i, head in enumerate(heads):
    heads[i] = heads[i] - i
```

**Comment by adrianeboyd** [comment]
I can do it. Maybe it's a chance to figure out some of the old bugs with `set_children_from_heads()`, too.

**Comment by svlandeg** [comment]
Fixed by PR #5079, which now throws an error pointing to the fact that the heads need to be relative, in this specific case:
```
heads = [1, 2, 1, 0, 1, -2, 1, -2, -1, -1, -1, -2, -3, -1]
```
